### PR TITLE
[JIMT][LABS-235] prevent large data tables from crashing the app

### PR DESCRIPTION
**Pull request by thomasoniii**
There are two separate zen desk tickets that came in. They reference different tables, but the underlying issue is the same - selecting certain data sets will crash the app.

There are two separate issues - the first one is easy to fix, and it was hiding the second, which was trickier. The second issue is resolved, but we should have discussion around the arbitrary limits on it and potentially styling.

The underlying issue is simply that these tables have too much data in them properly to visualize in a cross-product view like this.

First off, the initial problem was in `CrossTabChart` where we were calculating the min/max values. That was done by spreading the `numericValues` array and handing it into the `min` and `max` functions. But these arrays were too big and there are limits on the number of arguments a function can receive (it seems to be implementation specific). This is fixed by replacing the spread calls with a single `reduce`. It has the advantage of also only looping over the array once to get min and max, instead of looping once for each.

But that was only hiding the next problem - rendering too much data. For example, zen desk 465096 suggested an x value of `name`, and that anything in the second column would crash it.  In truth (once the min/max -> reduce change was made), only certain values would crash it. For example, `name x id` would crash but `name x year released` would not. This was due to the number of cells being rendered to the screen - upwards of 250,000 of them in the bad cases, and "only" 50,000 or so in the "good" cases. This is simply too much data.

As one option, we could virtualize the table - there's a partial reference implementation of this in a closed PR here: https://github.com/code-dot-org/code-dot-org/pull/55624

But that adds a lot of complexity to the app, and the reality is that no one should really be attempting to visualize that much data any way - it's too easy to get lost in the table, and it's of dubious value. So instead of that approach, we're preventing the user from doing it entirely.

To that end, three constants were added to the CrossTabChart:

```
export const MAX_CROSSTAB_COLUMNS = 150;
export const MAX_CROSSTAB_ROWS = 10000;
export const MAX_CROSSTAB_CELLS = 10000;
```

Those mean that a cross tab chart can have at most 150 columns, 10000 rows, and 10000 cells. Meaning you can have a table with 1 column and 10,000 rows and be fine. But if you have a table with 2 columns and 10,000 rows, you'll exceed the cell count. Likewise, things like 20 columns and 500 rows are fine. or 2 columns and 5,000 rows. 5,000 columns of 2 rows would not be allowed.

**Please note these values are arbitrary**. I picked reasonable sounding values, but we can crank them up higher if desired. The higher they go, the more likely the table will crash. Tables with high row and column counts are also extremely difficult to navigate w/o virtualized sticky headers.

To prevent it entirely, if a user attempts to create a cross tab chart, they will only be allowed to select a column if it will have fewer than the max number of columns, as such:

Screenshot 2024-01-11 at 12 01 41 PM

When it comes to selecting the row, 
Screenshot 2024-01-11 at 12 01 51 PM

Values will be filtered so charts can only be created within the row and cell counts.

In this case, as of today, `board games.year released` has `106` entries. So it's under the 150 column count, but 106 rows x 106 columns would be 11,236 total cells and exceed the cell count, even though there are only 106 rows. If you selected a different first value (e.g, minimum players), you would be able to select `year released` for the row.

You can get around this with filters:

Screenshot 2024-01-11 at 12 02 05 PM

In this case, since we have filtered to only games released in `1977`, we no longer have the excessive amount of data, so all of the columns are available. There may be limitations on the rows available, depending upon the cross product.

If the user selects a column which cannot be combined with _any_ row, then the second drop box will simply be empty and they will not be able to proceed.

Further, the chart itself is protected - it also looks to the max column/row/cell values and won't display if they're exceeded.

Screenshot 2024-01-11 at 12 03 03 PM
Screenshot 2024-01-11 at 12 03 36 PM
Screenshot 2024-01-11 at 12 03 44 PM

These error messages should not be displayed to the user, because they should be prevented from creating a chart in an invalid state to begin with. This is around in case an existing chart is loaded, or if a chart with existing data is later increased to the point that it might crash.

This PR should be merged in ASAP to prevent the crashes. That said, there are a couple of points to consider in the short term as well as the long term:

Short Term Points
- are those arbitrary limits acceptable? Should they be rasied/lowered?
- The messages for the crash conditions are terse and not pretty and not localized. I don't think we should care enough to localize right now and should defer, but maybe it's important enough.

Longer term points:
- All of these components should be refactored. This is bringing in even more logic specific to CrossTab chart into `VIsualizerModal` and it should be broken up.
- CrossTab really should be a virtualized table.
- Should we have better instruction/error reporting/navigability for the user to help them figure out why columns aren't included or that they should include filters to choose things? I think these are necessary additions, but not at the expense of letting the crashes linger until they're decided upon.

## Links

- jira ticket: [LABS-235](https://codedotorg.atlassian.net/browse/LABS-235)
- Zendesk: [465096](https://codeorg.zendesk.com/agent/tickets/465096)
- Zendesk: [475504](https://codeorg.zendesk.com/agent/tickets/475504)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: This function will iterate over the cross product and count the number of values for all of the headers.

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: And this one will filter out our table columns based upon whether they exceed our thresholds.

If there are no countedCells, we filter nothing. If there is a `selectedColumn1`, then we also filter on the cross product to count the number of cells.

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (bencodeorg) [inline apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:55]: Maybe a comment here covering why localization isn't necessary (these shouldn't be shown to users)?

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (bencodeorg) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: Is what this returns more like "buckets" (plural)? How can I envision what a "bucket" represents?

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (bencodeorg) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: Can we disable the dropdown if there are no options to choose from?

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (bencodeorg) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: Having a little trouble following this, is it producing a count of values? or a set of distinct values per header (which then is counted elsewhere)?

**Review by bencodeorg — COMMENTED** [review]
A couple clarification questions for my understanding (unfamiliar code 😬 ), but overall approach looks reasonable to me!

**Comment by bencodeorg** [comment]
> Should we have better instruction/error reporting/navigability for the user to help them figure out why columns aren't included or that they should include filters to choose things? I think these are necessary additions, but not at the expense of letting the crashes linger until they're decided upon.

It does seem like it might be valuable to have better messaging here about why columns are available vs not, could you log a follow-up in Jira (maybe soliciting input from Sam)?

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (fisher-alice) [inline apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:10]: Thanks for checking with curriculum already on these limits!

Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:10]: Btw - this tool is used in [this CSP lesson](https://studio.code.org/s/csp5-2023/lessons/4) and in the [teaching slides](https://docs.google.com/presentation/d/14G7ptwES6IcNnpXccTD-09noprIzbdnXuYM3aVEeFJE/edit#slide=id.g206826bc1d5_0_309) that discuss the usefulness of 'cross tab' charts, they state 'Not useful: If either column has too many values (the chart would be enormous)'!

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: Maybe we could name function something like `getDataColumnBuckets` and assign it to `dataColumnBuckets`? And in a comment, you could add that a single bucket is the set of unique values for a particular column?
As you've said before about hardest things in CS...

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: If you take a look at scatter plot visualization, columns with numeric values are displayed while non-numeric columns are still displayed but shown as disabled. 

Screenshot 2024-01-12 at 4 55 42 PM

This then lets users know that these columns are intentionally disabled. 

I wonder about doing something similar for cross tab tables so that 
1) we're consistent with how other tables are filtering columns
2) users may be alerted to how this is intentional
3) it's nice to know which columns were disabled!

**Review by fisher-alice — COMMENTED** [review]
Awesome work! I left a couple questions for you. I appreciate the thorough explanations and investigation into the option of using virtualized tables - learned a lot from reading about it!

I suggested looking at how scatter plots handle the filtering of columns. +1 to Ben's comment about needing to message to users why columns are not included or enabled. Thanks!

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: But as Ben mentioned already, we'd need to include a way to let users know why these columns are disabled. Maybe we could add an explanation message below the x values and y values dropdowns when columns are filtered.

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (sanchitmalhotra126) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:55]: If these aren't user facing, would it be more useful to log these to our metrics system? What situation would cause these to appear?

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: _Technically_ it's producing a `Set`, but then we just read the `size` of the set to get the count of unique values. No further calculations required.

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:55]: These are _potentially_ user facing, but they're just an escape hatch in case a user has managed to somehow bypass the filters on which columns they can select and produced a chart that'd crash.

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: I'd really rather leave this as `bucket`, since I write _all_ my `reduce` functions with `bucket` as the first argument.

It's just the accumulator used by the reduce function - so it's never actually returned or referenced outside of the function. It's just a known place to stick whatever the appropriate return value is, in whatever form it may need to be (object, array, summation, whatever), and any sort of renaming it to something more significant is going to hide the accumulator as well as attach additional external meaning to it.

We're not getting buckets, we're getting sets which we are _only_ using to get sizes of.  This could be wrapped into other code to re-map it to an object pointing from column -> count (instead of column -> set) but that's just adding an extra loop to save accessing a prop on an object later, so I'm not super in favor of it. It's extra complexity and runtime for minimal benefit.

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: I'm leaving it enabled, but I did flip around the logic so that it'll pop up and show the columns as disabled, potentially up to all of them.

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:10]: Yes...but...it actually _might_ be useful if we had a properly virtualized table with sticky column/row headers and search capability. 😉

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: I did flip around the logic so that it'll keep the columns available, but disabled.

But I don't want to add in messages yet - we already don't display messages for non-numeric columns being disabled on other charts, so any messaging should be rolled into a bigger effort for consistent messaging across charts. Good follow up task!

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:55]: And localization _might_ be desired at some point, but we also may want to completely re-do the view, style it, change the behavior, etc. So right now I'm just deferring it as a follow up task for how we want to actually present the error case and then we can try localizing that.

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:10]: That would be cool! Would be interested in knowing what curriculum folks thought about this (if worth the investment).

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: Thanks! And just clarifying that my suggestion was to rename the function `countCells` to `getDataColumnBuckets`. I agree `bucket` makes sense here.

Comment (bencodeorg) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: Do you mean you use the name "bucket" for the accumulator any time you use a reducer? I see another reduce function that uses the name "values" above :)

Fair enough if you want to keep it mapped column -> set to avoid the extra loop, but I think it'd be clearer if the variable name here reflected that (`getValuesByColumn` (?) instead of `countCells`)

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: From a teaching perspective, I think:
- disabling non-numeric columns in histograms
- disabling columns that result in cell count greater than an arbitrarily set constant in cross-tab charts  

are two different cases.

In the lesson plan, it is explicitly stated that 'Histograms can only be created with numerica data...'  [lesson_slides](https://docs.google.com/presentation/d/1jrF77jto1ZR_XWchlm0DHNxLygkx5L-FQQlqr9VYTU8/edit#slide=id.g2067e88de27_0_398) | [lesson_plan](https://studio.code.org/s/csp5-2023/lessons/2)

Screenshot of lesson plan:
Screenshot 2024-01-18 at 12 24 30 PM

But there isn't any mention in the lesson plan of limiting the columns based on size of dataset.

Maybe ask curriculum? I defer to them on this. Thanks!

File: apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx:55]: I added localization, and arbitrarily tossed the variables into the `common` file, since it looked like the best fit atm.

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx:205]: Wow. So I did. That might be a first. 😆 

I did rename it to `getValuesByColumn`, and the `countedCells` variable to `valuesByColumn`

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (thomasoniii) [reply apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: I'll file this as a future work item, there's too much UI consideration here to figure out how/when to display an error message, but I don't want to keep holding up this PR because it's still breaking in production.

Filed as https://codedotorg.atlassian.net/browse/LABS-462

File: apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx
Comment (fisher-alice) [inline apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx]: What do you think of renaming [`disabledOptions`](https://github.com/code-dot-org/code-dot-org/blob/0b09e235e21aef0843519e53aee09fc5d568ebc9/apps/src/storage/dataBrowser/dataVisualizer/VisualizerModal.jsx#L251) to `disabledNonNumericOptions` to distinguish from disabled options that are too big?

**Review by fisher-alice — APPROVED** [review]
Thanks for iterating on this! Definitely encourage a fast follow-up to communicate to users why columns are disabled. 
Nice solution!

**Review by bencodeorg — APPROVED** [review]
LGTM!

**Pull request merged**
